Add a tool to gather safer C++ statistics per project

rniwa opened the pull request:
#### 6f763fdcae30f6a69ed587b9006d9509d6062684
<pre>
Add a tool to gather safer C++ statistics per project
<a href="https://bugs.webkit.org/show_bug.cgi?id=297372">https://bugs.webkit.org/show_bug.cgi?id=297372</a>

Reviewed by Brianna Fan.

Add compute-safer-cpp-statistics to gather safer C++ statistics.
The total reports the total pass rate across all checkers as well as pass rate per checker.

For example, as of this writing, we gather the following statistics for Source/WebKit:
Overall: 79.70% safe
ForwardDeclChecker: 99.74% safe
MemoryUnsafeCastChecker: 99.74% safe
NoUncheckedPtrMemberChecker: 100.00% safe
NoUncountedMemberChecker: 99.97% safe
NoUnretainedMemberChecker: 97.86% safe
RefCntblBaseVirtualDtor: 100.00% safe
RetainPtrCtorAdoptChecker: 97.35% safe
UncheckedCallArgsChecker: 97.37% safe
UncheckedLocalVarsChecker: 99.91% safe
UncountedCallArgsChecker: 89.88% safe
UncountedLambdaCapturesChecker: 99.92% safe
UncountedLocalVarsChecker: 100.00% safe
UnretainedCallArgsChecker: 91.14% safe
UnretainedLambdaCapturesChecker: 100.00% safe
UnretainedLocalVarsChecker: 100.00% safe

* Tools/Scripts/compute-safer-cpp-statistics: Added.
(parser):
(matching_checkers):
(load_expectations):
(count_number_of_lines):
(enumerate_relevant_files):
(gather_statistics):
(safe_ratio):
(print_percentage):
(main):
* Tools/Scripts/webkitpy/safer_cpp/checkers.py:
(Checker.project_path):
(Checker):
(Checker.derived_sources_path):

Canonical link: <a href="https://commits.webkit.org/298668@main">https://commits.webkit.org/298668@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cb0473acc6de396b0ab35398a9c5baee2651f1ac

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116233 "1 style error") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/131/builds/35894 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26440 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/122290 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/66793 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/bf90e82f-191a-4a71-9f8c-45d132b36388) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/118122 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/36588 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44482 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/122290 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/66793 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/e0b24967-2f13-47da-ba4c-9c8b7772c821) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119182 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/29191 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104285 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/122290 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/115607 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/28291 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22394 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/65971 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/98577 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22540 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/125439 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/43127 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/32385 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/125439 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/43492 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100484 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/125439 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/19977 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/39074 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18570 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/43014 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/42481 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/45816 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/44185 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/0fa6760208da2a3c2a07f36a4b8e5ed86301ce3c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49366

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/84f50249d55733fabe727468bebde8b95ebd5af9. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49366

briannafan reviewed (approved):
Approved, with minor edits

rniwa commented on Tools/Scripts/compute-safer-cpp-statistics:
```diff
@@ -0,0 +1,172 @@
+        help='Specify the directory to ignore relative to the project root (e.g. WebProcess for Source/WebKit/WebProcess)'
+    )
+    parser.add_argument(
+        '--csv', '-v',
```
Oh sure, good point.

rniwa commented on Tools/Scripts/compute-safer-cpp-statistics:
```diff
@@ -0,0 +1,172 @@
+
+def count_number_of_lines(file_path):
+    with open(file_path) as file_handle:
+        return sum([1 for _ in file_handle])
```
Sure, removed.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/cb0473acc6de396b0ab35398a9c5baee2651f1ac. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49366

rniwa committed:
Add a tool to gather safer C++ statistics per project
https://bugs.webkit.org/show_bug.cgi?id=297372

Reviewed by Brianna Fan.

Add compute-safer-cpp-statistics to gather safer C++ statistics.
The total reports the total pass rate across all checkers as well as pass rate per checker.

For example, as of this writing, we gather the following statistics for Source/WebKit:
Overall: 79.70% safe
ForwardDeclChecker: 99.74% safe
MemoryUnsafeCastChecker: 99.74% safe
NoUncheckedPtrMemberChecker: 100.00% safe
NoUncountedMemberChecker: 99.97% safe
NoUnretainedMemberChecker: 97.86% safe
RefCntblBaseVirtualDtor: 100.00% safe
RetainPtrCtorAdoptChecker: 97.35% safe
UncheckedCallArgsChecker: 97.37% safe
UncheckedLocalVarsChecker: 99.91% safe
UncountedCallArgsChecker: 89.88% safe
UncountedLambdaCapturesChecker: 99.92% safe
UncountedLocalVarsChecker: 100.00% safe
UnretainedCallArgsChecker: 91.14% safe
UnretainedLambdaCapturesChecker: 100.00% safe
UnretainedLocalVarsChecker: 100.00% safe

* Tools/Scripts/compute-safer-cpp-statistics: Added.
(parser):
(matching_checkers):
(load_expectations):
(count_number_of_lines):
(enumerate_relevant_files):
(gather_statistics):
(safe_ratio):
(print_percentage):
(main):
* Tools/Scripts/webkitpy/safer_cpp/checkers.py:
(Checker.project_path):
(Checker):
(Checker.derived_sources_path):

Canonical link: https://commits.webkit.org/298668@main
